[READY] Improved performance of `Period`'s default formatter (`period_format`)

smarie opened the pull request:
`Period`'s default formatter (`period_format`) is now significantly (~twice) faster. 

- This improves performance of `str(Period)`, `repr(Period)`, and `Period.strftime(fmt=None)`, 
- as well as `PeriodArray.strftime(fmt=None)`, `PeriodIndex.strftime(fmt=None)` and `PeriodIndex.format(fmt=None)`. 
- Finally, `to_csv` operations involving `PeriodArray` or `PeriodIndex` with default `date_format` are also significantly accelerated.

ASVs:

```
> asv compare d82f9ddb c1bbfd6c
       before           after         ratio
     [d82f9ddb]       [c1bbfd6c]
-     3.12±0.05μs         998±40ns     0.32  tslibs.period.PeriodUnaryMethods.time_repr('M')
-     3.31±0.05μs       1.58±0.1μs     0.48  tslibs.period.PeriodUnaryMethods.time_repr('min')
-     3.15±0.09μs      1.05±0.07μs     0.33  tslibs.period.PeriodUnaryMethods.time_str('M')
-     3.20±0.05μs       1.75±0.1μs     0.55  tslibs.period.PeriodUnaryMethods.time_str('min')
-      2.90±0.1μs         850±40ns     0.29  tslibs.period.PeriodUnaryMethods.time_strftime_default('M')
-     3.01±0.01μs       1.48±0.1μs     0.49  tslibs.period.PeriodUnaryMethods.time_strftime_default('min')
-     4.16±0.05ms      2.37±0.08ms     0.57  strftime.PeriodStrftime.time_frame_period_formatting_default(1000, 'D')
-     4.35±0.06ms       2.63±0.1ms     0.60  strftime.PeriodStrftime.time_frame_period_formatting_default(1000, 'H')
-      40.3±0.6ms       21.8±0.2ms     0.54  strftime.PeriodStrftime.time_frame_period_formatting_default(10000, 'D')
-      43.5±0.2ms         26.3±1ms     0.60  strftime.PeriodStrftime.time_frame_period_formatting_default(10000, 'H')
      4.58±0.08ms      2.88±0.07ms    ~0.63  strftime.PeriodStrftime.time_frame_period_formatting_index_default(1000, 'D')
       4.73±0.2ms      3.02±0.05ms    ~0.64  strftime.PeriodStrftime.time_frame_period_formatting_index_default(1000, 'H')
-      42.8±0.3ms       24.0±0.5ms     0.56  strftime.PeriodStrftime.time_frame_period_formatting_index_default(10000, 'D')
-      43.2±0.3ms       27.3±0.3ms     0.63  strftime.PeriodStrftime.time_frame_period_formatting_index_default(10000, 'H')
-     4.16±0.05ms       2.41±0.3ms     0.58  strftime.PeriodStrftime.time_frame_period_to_str(1000, 'D')
-     4.28±0.03ms       2.71±0.1ms     0.63  strftime.PeriodStrftime.time_frame_period_to_str(1000, 'H')
-      41.6±0.7ms       22.7±0.2ms     0.55  strftime.PeriodStrftime.time_frame_period_to_str(10000, 'D')
-      42.9±0.6ms       26.0±0.5ms     0.61  strftime.PeriodStrftime.time_frame_period_to_str(10000, 'H')
```
EDIT: additional for `to_csv`
```
-     5.94±0.05ms       3.94±0.3ms     0.66  io.csv.ToCSVPeriod.time_frame_period_formatting(1000, 'D') 
-      6.39±0.1ms      4.37±0.04ms     0.68  io.csv.ToCSVPeriod.time_frame_period_formatting(1000, 'H') 
-      51.8±0.2ms       30.2±0.3ms     0.58  io.csv.ToCSVPeriod.time_frame_period_formatting(10000, 'D')
-       59.8±20ms       35.0±0.2ms     0.59  io.csv.ToCSVPeriod.time_frame_period_formatting(10000, 'H')
-      7.09±0.2ms       3.94±0.2ms     0.56  io.csv.ToCSVPeriod.time_frame_period_no_format(1000, 'D')  
-      6.72±0.1ms      4.30±0.03ms     0.64  io.csv.ToCSVPeriod.time_frame_period_no_format(1000, 'H')  
-        54.0±2ms       30.5±0.3ms     0.56  io.csv.ToCSVPeriod.time_frame_period_no_format(10000, 'D') 
-        54.4±1ms       35.3±0.5ms     0.65  io.csv.ToCSVPeriod.time_frame_period_no_format(10000, 'H')
       9.94±0.3ms       9.70±0.1ms     0.98  io.csv.ToCSVPeriodIndex.time_frame_period_formatting_index(1000, 'D')
      9.69±0.07ms       9.84±0.7ms     1.02  io.csv.ToCSVPeriodIndex.time_frame_period_formatting_index(1000, 'H')
       85.4±0.9ms       84.9±0.9ms     0.99  io.csv.ToCSVPeriodIndex.time_frame_period_formatting_index(10000, 'D')
         89.3±2ms         87.3±3ms     0.98  io.csv.ToCSVPeriodIndex.time_frame_period_formatting_index(10000, 'H')
-      6.32±0.3ms      3.78±0.03ms     0.60  io.csv.ToCSVPeriodIndex.time_frame_period_no_format_index(1000, 'D')
-      6.30±0.2ms       4.60±0.4ms     0.73  io.csv.ToCSVPeriodIndex.time_frame_period_no_format_index(1000, 'H')
         50.9±2ms         33.9±8ms    ~0.66  io.csv.ToCSVPeriodIndex.time_frame_period_no_format_index(10000, 'D')
         52.6±1ms         35.6±6ms    ~0.68  io.csv.ToCSVPeriodIndex.time_frame_period_no_format_index(10000, 'H')
```

- [x] closes one part of #44764 
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Commit:
Improved performance of default period formatting (`period_format`). Added corresponding ASVs

Commit:
Improved ASV for period frames and datetimes

Commit:
What's new

smarie committed:
Update asv_bench/benchmarks/strftime.py

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

Commit:
Fixed whats new backticks

Commit:
Merge branch 'feature/44764_perf_issue_new_period' of https://github.com/smarie/pandas into feature/44764_perf_issue_new_period

Commit:
Completed whatsnew

smarie commented:
It seems that the failing tests are not related to this PR's contents, do you confirm ?

Commit:
Added ASVs for to_csv for period

Commit:
Aligned the namings

Commit:
Completed Whats new

smarie commented:
I added a few asvs for the to_csv operations and updated the what's new. This is now ready for review @MarcoGorelli

Commit:
Added a docstring explaining why the ASV bench with custom date format was improved: the date_format parameter is not taken into account today.

smarie commented:
Note: the failing test seeems to be related to a network issue, so completely unrelated. 

> FAILED pandas/tests/io/parser/common/test_file_buffer_url.py::test_file_descriptor_leak[c_high] - AssertionError: ([], [pconn(fd=14, family=<AddressFamily.AF_INET6: 10>, type=<SocketKind.SOCK_STREAM: 1>, laddr=addr(ip='::1', port=35970), raddr=addr(ip='::1', port=3306), status='ESTABLISHED')])

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

MarcoGorelli commented on asv_bench/benchmarks/strftime.py:
```diff
@@ -45,6 +48,44 @@ def time_frame_datetime_formatting_custom(self, obs):
+    def time_frame_period_formatting_index_default(self, obs, fq):
+        self.data.set_index("p").index.format()
+
+    def time_frame_period_formatting_custom(self, obs, fq):
+        self.data["p"].dt.strftime(date_format="%Y-%m-%d --- %H:%M:%S")
+
+    def time_frame_period_formatting_iso8601_strftime_Z(self, obs, fq):
+        self.data["p"].dt.strftime(date_format="%Y-%m-%dT%H:%M:%SZ")
+
+    def time_frame_period_formatting_iso8601_strftime_offset(self, obs, fq):
+        """Not optimized yet as %z is not supported by `convert_strftime_format`"""
+        self.data["p"].dt.strftime(date_format="%Y-%m-%dT%H:%M:%S%z")
```
why do these not show up in the asv results you posted?

smarie replied:
good point, it seems that I forgot them in the copy paste. I'll try to rerun asv partially to get them

smarie replied:
EDIT: Actually I had removed all asvs that were expected not to show improvements (custom date formats). Here are all ASVs for `PeriodStrftime`:

```
       15.4±0.2ms         17.1±2ms    ~1.11  strftime.PeriodStrftime.time_frame_period_formatting_custom(1000, 'D')
       15.0±0.7ms       14.1±0.2ms     0.94  strftime.PeriodStrftime.time_frame_period_formatting_custom(1000, 'H')
          142±6ms          153±5ms     1.08  strftime.PeriodStrftime.time_frame_period_formatting_custom(10000, 'D')
          143±5ms          157±7ms     1.10  strftime.PeriodStrftime.time_frame_period_formatting_custom(10000, 'H')
-     4.37±0.07ms      2.24±0.06ms     0.51  strftime.PeriodStrftime.time_frame_period_formatting_default(1000, 'D')
-     4.72±0.06ms       2.90±0.1ms     0.61  strftime.PeriodStrftime.time_frame_period_formatting_default(1000, 'H')
-        42.8±1ms       24.1±0.7ms     0.56  strftime.PeriodStrftime.time_frame_period_formatting_default(10000, 'D')
-      48.4±0.5ms       24.4±0.2ms     0.50  strftime.PeriodStrftime.time_frame_period_formatting_index_default(10000, 'D')
-        50.9±3ms       27.1±0.7ms     0.53  strftime.PeriodStrftime.time_frame_period_formatting_index_default(10000, 'H')
       7.52±0.7ms       7.22±0.3ms     0.96  strftime.PeriodStrftime.time_frame_period_formatting_iso8601_strftime_Z(1000, 'D')
       7.12±0.1ms       7.00±0.3ms     0.98  strftime.PeriodStrftime.time_frame_period_formatting_iso8601_strftime_Z(1000, 'H')
         75.2±4ms         68.1±1ms    ~0.91  strftime.PeriodStrftime.time_frame_period_formatting_iso8601_strftime_Z(10000, 'D')
         71.9±4ms         77.3±8ms     1.08  strftime.PeriodStrftime.time_frame_period_formatting_iso8601_strftime_Z(10000, 'H')
       13.5±0.5ms       14.0±0.4ms     1.03  strftime.PeriodStrftime.time_frame_period_formatting_iso8601_strftime_offset(1000, 'D')
         15.9±3ms         14.4±1ms    ~0.91  strftime.PeriodStrftime.time_frame_period_formatting_iso8601_strftime_offset(1000, 'H')
          135±3ms          148±7ms     1.10  strftime.PeriodStrftime.time_frame_period_formatting_iso8601_strftime_offset(10000, 'D')
          140±3ms          136±2ms     0.97  strftime.PeriodStrftime.time_frame_period_formatting_iso8601_strftime_offset(10000, 'H')

```

MarcoGorelli reviewed (commented):
Nice one, the changes themselves look good to me! Just got a question

Regarding `date_format` only taking effect for `Period` if it's in the index - that looks like a bug to me, well done for noticing! I don't see any issue open about this, do you want to open one? (else I can, no worries)

smarie commented:
Thanks @MarcoGorelli !

> Regarding date_format only taking effect for Period if it's in the index - that looks like a bug to me, well done for noticing! I don't see any issue open about this, do you want to open one? (else I can, no worries)

Will do, ok

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

# Conflicts:
#	doc/source/whatsnew/v1.5.4.rst

Commit:
Moved whatsnew to 2.0.0

MarcoGorelli reviewed (approved):
Looks good to me, thanks @smarie !

Leaving open for a bit in case others have comments

smarie commented:
Thanks @MarcoGorelli . Any reason this is for 2.1 and not for 2.0 ? If so I need to move the what's new lines  :)

Commit:
Moved whatsnew to 2.1

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

smarie commented:
Ok, I moved the whatsnew.

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

# Conflicts:
#	doc/source/whatsnew/v2.1.0.rst

MarcoGorelli commented:
sure - there's been no objections in several days, so I'll merge on green

thanks for your work here!

smarie commented:
Thanks @MarcoGorelli ! Unfortunately there is one red, completely unrelated (numpy https://github.com/pandas-dev/pandas/actions/runs/4324702806/jobs/7549886360#step:8:287 )

`pandas/tests/apply/test_frame_apply_relabeling.py::test_agg_relabel_partial_functions - ValueError: cannot reindex on an axis with duplicate labels`

So you can either proceed "as is" or wait until someone fixes the issue on the main branch

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new_period

smarie commented:
Confirmed: the "red" on CI is described in #51782, totally unrelated

jbrockmendel commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1152,46 +1152,89 @@ cdef int64_t period_ordinal_to_dt64(int64_t ordinal, int freq) except? -1:
-    return _period_strftime(value, freq, fmt)
+
+cdef str _period_default_format(int64_t value, int freq):
+    """A faster default formatting function leveraging string formatting."""
```
faster than what?  i assume this refers to the implementation this replaces, but that wont be clear to future readers

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

# Conflicts:
#	doc/source/whatsnew/v2.1.0.rst

Commit:
Improved docstring as per code review

Commit:
Merge remote-tracking branch 'origin/feature/44764_perf_issue_new_period' into feature/44764_perf_issue_new_period

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new_period

jbrockmendel commented on asv_bench/benchmarks/io/csv.py:
```diff
@@ -98,6 +99,44 @@ def time_frame_date_no_format_index(self):
+    fname = "__test__.csv"
+
+    params = ([1000, 10000], ["D", "H"])
+    param_names = ["obs", "fq"]
```
can these be "nobs" and "freq", would match patterns elsewhere

smarie replied:
Ok. I just originally recopied what was already there in `ToCSVDatetimeBig`. Changing all.

jbrockmendel commented on asv_bench/benchmarks/io/csv.py:
```diff
@@ -98,6 +99,44 @@ def time_frame_date_no_format_index(self):
+        self.data.to_csv(self.fname)
+
+    def time_frame_period_formatting(self, obs, fq):
+        """IMPORTANT: date_format is not actually taken into account here today
```
can you make this a comment instead of a docstring.

not sure what to make of IMPORTANT here.  i often preface comments with "Note:" or "NB:" depending on what i think the reader needs to know.

smarie replied:
ok. I tried something, let me know if it is better

jbrockmendel commented on asv_bench/benchmarks/strftime.py:
```diff
@@ -45,6 +48,44 @@ def time_frame_datetime_formatting_custom(self, obs):
+        self.data["p"].astype(str)
+
+    def time_frame_period_formatting_default(self, obs, fq):
+        """Note that as opposed to datetimes, the default format of periods are
```
docstring->comment pls

"Periods are many and depend from the period characteristics" -> "a Period depends on its 'freq'"

smarie replied:
ok

jbrockmendel commented on asv_bench/benchmarks/strftime.py:
```diff
@@ -45,6 +48,44 @@ def time_frame_datetime_formatting_custom(self, obs):
+        self.data["p"].dt.strftime(date_format=None)
+
+    def time_frame_period_formatting_index_default(self, obs, fq):
+        self.data.set_index("p").index.format()
```
this is going to conflate the timing of format with set_index.  is there a viable alternative?

smarie replied:
ok fixed

Commit:
Renamed asv params as per code review

Commit:
Fixed ASV comment as per code review

Commit:
ASV: renamed parameters as per code review

Commit:
Improved `period_format`: now the performance is the same when no format is provided and if an explicit `fmt` is provided matching the actual default format for this `freq`

Commit:
Code review: Improved strftime ASV: set_index is now in the setup

Commit:
Merge remote-tracking branch 'origin/feature/44764_perf_issue_new_period' into feature/44764_perf_issue_new_period

smarie commented:
I fixed everything but I realized that the explicit default format could be actually caught, so I reworked a bit the period_format method.

I therefore need to re-run the ASVs to be sure this is ok

Commit:
Removed useless main

Commit:
Removed wrong code

Commit:
Improved ASVs for period formatting: now there is a "default explicit" everywhere

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

smarie commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1152,46 +1152,77 @@ cdef int64_t period_ordinal_to_dt64(int64_t ordinal, int freq) except? -1:
+    get_date_info(value, freq, &dts)
+    freq_group = get_freq_group(freq)
+
+    # get the appropriate default format depending on frequency group
```
```suggestion
    # use the appropriate default format depending on frequency group
```

smarie committed:
Update pandas/_libs/tslibs/period.pyx

smarie committed:
Update pandas/_libs/tslibs/period.pyx

smarie commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1152,46 +1152,78 @@ cdef int64_t period_ordinal_to_dt64(int64_t ordinal, int freq) except? -1:
+                f".{((dts.us * 1000) + (dts.ps // 1000)):09d}")
+
+    elif fmt is None:
+        raise ValueError(f"Unknown freq: {freq}")
```
```suggestion
        # `freq_group` is invalid, raise
        raise ValueError(f"Unknown freq: {freq}")
```

smarie committed:
Update pandas/_libs/tslibs/period.pyx

smarie commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1152,46 +1152,79 @@ cdef int64_t period_ordinal_to_dt64(int64_t ordinal, int freq) except? -1:
+    freq_group = get_freq_group(freq)
+
+    # use the appropriate default format depending on frequency group
+    if freq_group == FR_ANN and (fmt is None or fmt == "%Y"):
```
I refactored all of this passage so that for each frequency group, both fmt=None and fmt=<the default fmt> are equally fast

Commit:
Minor refactoring to avoid retesting for none several time

Commit:
Merge branch 'feature/44764_perf_issue_new_period' of https://github.com/smarie/pandas into feature/44764_perf_issue_new_period

# Conflicts:
#	pandas/_libs/tslibs/period.pyx

Commit:
Fixed issue: bool does not exist, using bint

Commit:
Added missing quarter variable as cdef

Commit:
Fixed asv bug

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

# Conflicts:
#	doc/source/whatsnew/v2.1.0.rst

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

# Conflicts:
#	doc/source/whatsnew/v2.1.0.rst

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new_period

smarie commented:
@jbrockmendel @MarcoGorelli would you have a few mins this week to check this PR so that I can move to the next one ? 
or maybe @mroeschke  ?

MarcoGorelli commented:
yup taking a look now (was off last week)

MarcoGorelli commented on pandas/_libs/tslibs/period.pyx:
```diff
@@ -1155,46 +1155,81 @@ cdef int64_t period_ordinal_to_dt64(int64_t ordinal, int freq) except? -1:
+        return f"{dts.year}"
+
+    elif freq_group == FR_QTR and (is_fmt_none or fmt == "%FQ%q"):
+        # get quarter and modify dts.year to be the fiscal year (?)
```
the changes themselves look good, I'm just confused by this `(?)`. are you asking for confirmation?

smarie replied:
yes, "fiscal year" is the wording to confirm in this comment, I wasn't sure

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

# Conflicts:
#	doc/source/whatsnew/v2.1.0.rst

Commit:
Merge remote-tracking branch 'origin/feature/44764_perf_issue_new_period' into feature/44764_perf_issue_new_period

Commit:
Code review: fixed docstring

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

MarcoGorelli commented:
ok should be good - I'll give it a final look and merge this week

smarie committed:
Merge branch 'main' into feature/44764_perf_issue_new_period

MarcoGorelli reviewed (changes requested):
I've had another look, and to be honest I'm slightly scared about merging something this user-facing

What I'm most concerned about is that not all these reprs seems to be explicitly tested for. Some of them are, e.g.

https://github.com/pandas-dev/pandas/blob/044227f25bfbf61a9b7833c15538f8af26bfb89d/pandas/tests/scalar/period/test_period.py#L711-L719

but then `FR_SEC` isn't (as far as I can tell).

Reckon we should add such unit tests for each period group (perhaps in a separate PR, which we could get in very quickly)? That would give me much more confidence about merging this

Really sorry this is taking so long

smarie commented:
> Reckon we should add such unit tests for each period group (perhaps in a separate PR, which we could get in very quickly)? That would give me much more confidence about merging this

I did not review the available tests extensively so maybe this is indeed needed. I'll have a look one of these days

smarie commented:
Yes this is still active and even ready for merge, but it would require that another PR pushes a few more tests first so that @MarcoGorelli has more confidence :)

Commit:
Merge branch 'main' of https://github.com/pandas-dev/pandas into feature/44764_perf_issue_new_period

# Conflicts:
#	doc/source/whatsnew/v2.1.0.rst

smarie commented:
Now that #53003 is merged this should be ok - all new tests seem to pass.

MarcoGorelli reviewed (approved):
cool, let's get this in, thanks @smarie !

![](https://i.fluffy.cc/7WXqprWzDtCpbfprBZG1Hngnp7H7pfQm.gif)

Commit:
fixup

MarcoGorelli commented on doc/source/whatsnew/v2.1.0.rst:
```diff
@@ -291,7 +291,12 @@ Performance improvements
 - Performance improvement in :meth:`Series.to_numpy` when dtype is a numpy float dtype and ``na_value`` is ``np.nan`` (:issue:`52430`)
 - Performance improvement in :meth:`~arrays.ArrowExtensionArray.to_numpy` (:issue:`52525`)
--
+- :class:`Period`'s default formatter (`period_format`) is now significantly
+  (~twice) faster. This improves performance of ``str(Period)``, ``repr(Period)``, and
+  :meth:`Period.strftime(fmt=None)`, as well as ``PeriodArray.strftime(fmt=None)``,
+  ``PeriodIndex.strftime(fmt=None)`` and ``PeriodIndex.format(fmt=None)``. Finally,
+  ``to_csv`` operations involving :class:`PeriodArray` or :class:`PeriodIndex` with
+  default ``date_format`` are also significantly accelerated. (:issue:`51459`)
```
```suggestion
- :class:`Period`'s default formatter (`period_format`) is now significantly (~twice) faster. This improves performance of ``str(Period)``, ``repr(Period)``, and :meth:`Period.strftime(fmt=None)`, as well as ``PeriodArray.strftime(fmt=None)``, ``PeriodIndex.strftime(fmt=None)`` and ``PeriodIndex.format(fmt=None)``. Finally, ``to_csv`` operations involving :class:`PeriodArray` or :class:`PeriodIndex` with default ``date_format`` are also significantly accelerated. (:issue:`51459`)
```

MarcoGorelli replied:
single-line please

MarcoGorelli committed:
Update doc/source/whatsnew/v2.1.0.rst

MarcoGorelli commented:
let's get this in, thanks @smarie

smarie commented:
Great news @MarcoGorelli thanks ! 
I'll now finally move back to the main PR of this thread: #51298
Stay tuned ;)

Thanks again !
